polyfill/lib/datetime.mjs

```diff
@@ -231,32 +366,62 @@ export class DateTime {
       disambiguation
     ));
     const Construct = ES.SpeciesConstructor(this, DateTime);
-    const result = new Construct(year, month, day, hour, minute, second, millisecond, microsecond, nanosecond);
+    const result = new Construct(
+      year,
+      month,
+      day,
+      hour,
+      minute,
+      second,
+      millisecond,
+      microsecond,
+      nanosecond,
+      calendar
+    );
     if (!ES.IsTemporalDateTime(result)) throw new TypeError('invalid result');
     return result;
   }
   difference(other, options) {
     if (!ES.IsTemporalDateTime(this)) throw new TypeError('invalid receiver');
     if (!ES.IsTemporalDateTime(other)) throw new TypeError('invalid DateTime object');
+    const calendar = GetSlot(this, CALENDAR);
+    if (calendar.id !== GetSlot(other, CALENDAR).id) {
+      other = new DateTime(
+        GetSlot(other, ISO_YEAR),
+        GetSlot(other, ISO_MONTH),
+        GetSlot(other, ISO_DAY),
+        GetSlot(other, HOUR),
+        GetSlot(other, MINUTE),
+        GetSlot(other, SECOND),
+        GetSlot(other, MILLISECOND),
+        GetSlot(other, MICROSECOND),
+        GetSlot(other, NANOSECOND),
+        calendar
+      );
+    }
     const largestUnit = ES.ToLargestTemporalUnit(options, 'days');
     const [smaller, larger] = [this, other].sort(DateTime.compare);
     let { deltaDays, hours, minutes, seconds, milliseconds, microseconds, nanoseconds } = ES.DifferenceTime(
       smaller,
       larger
     );
-    let { year, month, day } = larger;
-    day += deltaDays;
+    let year = GetSlot(larger, ISO_YEAR);
+    let month = GetSlot(larger, ISO_MONTH);
+    let day = GetSlot(larger, ISO_DAY) + deltaDays;
     ({ year, month, day } = ES.BalanceDate(year, month, day));
 
+    const Date = GetIntrinsic('%Temporal.Date%');
+    const adjustedLarger = new Date(year, month, day, GetSlot(larger, CALENDAR));
     let dateLargestUnit = 'days';
     if (largestUnit === 'years' || largestUnit === 'months' || largestUnit === 'weeks') {
       dateLargestUnit = largestUnit;
     }
+    const dateOptions = ObjectAssign({}, options, { largestUnit: dateLargestUnit });
+    const dateDifference = calendar.difference(smaller, adjustedLarger, dateOptions);
 
-    let { years, months, weeks, days } = ES.DifferenceDate(smaller, { year, month, day }, dateLargestUnit);
-
+    let days;
     ({ days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds } = ES.BalanceDuration(
```

Comment:
Hmm.  What if the datetimes are 2020-05-01T01:00:00 and 2020-06-01T00:00:00, with an ISO calendar and largest date unit months.  Then you get 1 month on the first step and a residual -1 hours on the second step.  How do you resolve this?